File: examples/mini_game/sim.py
"""Simulation runner. Pure, deterministic, produces golden traces."""

from __future__ import annotations

from typing import List, Iterable, Dict, Any

from .core import Config, PlayerState, Event, Trace, TraceRecord
from .physics import update_player


def run_simulation(
    events: Iterable[Event],
    cfg: Config | None = None,
    initial: PlayerState | None = None,
) -> Trace:
    """Run the side-scroller simulation for a sequence of events.

    Returns a complete trace suitable for golden-master testing and behavior contracts.
    """
    cfg = cfg or Config()
    state = initial or PlayerState()

    # Index events by tick for O(1) lookup
    jump_ticks: set[int] = {e.tick for e in events if e.action == "jump"}

    trace: Trace = []

    for tick in range(cfg.max_ticks):
        did_jump = tick in jump_ticks
        update_player(state, did_jump=did_jump, cfg=cfg)

        rec = TraceRecord(
            tick=tick,
            x=round(state.x, 4),
            y=round(state.y, 4),
            vy=round(state.vy, 4),
            score=state.score,
            collided=state.collided,
        )
        trace.append(rec)

        if state.collided and tick > 5:  # allow a few steps after first collision for traces
            # continue to max_ticks so full behavior is captured
            pass

    return trace


def events_from_list(jump_ticks: List[int]) -> List[Event]:
    return [Event(tick=t, action="jump") for t in jump_ticks]


# Pre-defined input sequences for golden traces
GOLDEN_INPUTS: Dict[str, List[int]] = {
    "no_jumps": [],
    "jump_once": [3],
    "jump_to_clear_first": [3, 8],
    "aggressive": [1, 5, 12, 20, 28],
}

File: examples/mini_game_rust/src/sim.rs
//! Simulation runner, ported from mini_game/sim.py (structure-preserving).

use crate::core::{Config, Event, PlayerState, Trace, TraceRecord};
use crate::physics::update_player;

pub fn run_simulation(
    events: &[Event],
    cfg: Option<Config>,
    initial: Option<PlayerState>,
) -> Trace {
    let cfg = cfg.unwrap_or_default();
    let mut state = initial.unwrap_or_default();

    let jump_ticks: std::collections::HashSet<usize> = events
        .iter()
        .filter(|e| e.action == "jump")
        .map(|e| e.tick)
        .collect();

    let mut trace = Vec::new();

    for tick in 0..cfg.max_ticks {
        let did_jump = jump_ticks.contains(&tick);
        update_player(&mut state, did_jump, &cfg);

        let rec = TraceRecord {
            tick,
            x: (state.x * 10000.0).round() / 10000.0,
            y: (state.y * 10000.0).round() / 10000.0,
            vy: (state.vy * 10000.0).round() / 10000.0,
            score: state.score,
            collided: state.collided,
        };
        trace.push(rec);
    }

    trace
}

pub fn events_from_jumps(jumps: &[usize]) -> Vec<Event> {
    jumps
        .iter()
        .map(|&t| Event {
            tick: t,
            action: "jump".to_string(),
        })
        .collect()
}
